File: lab4-sistemas-operativos/go/dining_philosophers.go
/*
 * dining_philosophers.go
 *
 * Solución en Go al problema de los Filósofos Comensales.
 * Se emplean sync.Mutex para cada tenedor y un "camarero" implementado
 * con un canal (buffered) que solo permite N-1 filósofos intentando
 * comer simultáneamente.
 *
 * Compilar: go build dining_philosophers.go
 * Uso: ./dining_philosophers <num_philosophers> <num_ciclos_por_filosofo>
 */

package main

import (
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"sync"
	"time"
)

var numPhilosophers int
var cyclesPerPhilosopher int

// Cada tenedor es un mutex
type Fork struct {
	sync.Mutex
}

// Estructura del filósofo
type Philosopher struct {
	id       int
	left     int
	right    int
	forks    []Fork
	waiterCh chan struct{}
	wg       *sync.WaitGroup
}

// Filósofo piensa
func (p *Philosopher) think() {
	fmt.Printf("[Filósofo %d] Pensando...\n", p.id)
	time.Sleep(time.Duration(200+rand.Intn(200)) * time.Millisecond) // 200-400ms
}

// Filósofo come
func (p *Philosopher) eat(cycle int) {
	fmt.Printf("[Filósofo %d] Comiendo (ciclo %d)...\n", p.id, cycle)
	time.Sleep(time.Duration(250+rand.Intn(250)) * time.Millisecond) // 250-500ms
}

// Ciclo principal del filósofo
func (p *Philosopher) dine() {
	defer p.wg.Done()
	for i := 0; i < cyclesPerPhilosopher; i++ {
		p.think()

		// Solicitar permiso al camarero: si canal con buffer N-1 está vacío, bloquea
		p.waiterCh <- struct{}{}

		// Tomar tenedores en orden (menor índice primero)
		if p.left < p.right {
			p.forks[p.left].Lock()
			p.forks[p.right].Lock()
		} else {
			p.forks[p.right].Lock()
			p.forks[p.left].Lock()
		}

		p.eat(i)

		// Soltar tenedores
		p.forks[p.left].Unlock()
		p.forks[p.right].Unlock()

		// Liberar lugar en el camarero
		<-p.waiterCh
	}

	fmt.Printf("[Filósofo %d] Terminó todos sus ciclos.\n", p.id)
}

func main() {
	if len(os.Args) != 3 {
		fmt.Printf("Uso: %s <num_philosophers> <num_ciclos_por_filosofo>\n", os.Args[0])
		os.Exit(1)
	}
	numPhilosophers, _ = strconv.Atoi(os.Args[1])
	cyclesPerPhilosopher, _ = strconv.Atoi(os.Args[2])

	rand.Seed(time.Now().UnixNano())

	// Crear slice de tenedores
	forks := make([]Fork, numPhilosophers)

	// El canal del camarero tiene buffer de tamaño N-1
	waiterCh := make(chan struct{}, numPhilosophers-1)

	var wg sync.WaitGroup

	// Crear e iniciar filósofos
	for i := 0; i < numPhilosophers; i++ {
		p := &Philosopher{
			id:       i,
			left:     i,
			right:    (i + 1) % numPhilosophers,
			forks:    forks,
			waiterCh: waiterCh,
			wg:       &wg,
		}
		wg.Add(1)
		go p.dine()
	}

	wg.Wait()
	fmt.Println("Todos los filósofos han terminado.")
}

File: lab4-sistemas-operativos/c/dining_philosophers.c
/*
 * dining_philosophers.c
 *
 * Solución al problema de los Filósofos Comensales en C usando pthreads.
 * Se asegura que no haya deadlock usando un semáforo “camarero” que solo
 * permita a N-1 filósofos intentar tomar tenedores simultáneamente.
 *
 * Compilar: gcc dining_philosophers.c -o dining_philosophers -pthread -lrt
 * Uso: ./dining_philosophers <num_philosophers> <num_ciclos_por_filosofo>
 */

#include <pthread.h>
#include <semaphore.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

int num_philosophers;
int cycles_per_philosopher;

// Cada tenedor es un mutex
pthread_mutex_t *forks;

// Semáforo camarero (permite hasta num_philosophers-1 a la vez)
sem_t waiter;

typedef struct {
    int id;
} PhilosopherArgs;

// Simula pensar
void think(int id) {
    printf("[Filósofo %d] Pensando...\n", id);
    usleep(200000 + (rand() % 200000)); // 200-400 ms
}

// Simula comer
void eat(int id, int cycle) {
    printf("[Filósofo %d] Comiendo (ciclo %d)...\n", id, cycle);
    usleep(250000 + (rand() % 250000)); // 250-500 ms
}

void *philosopher(void *arg) {
    PhilosopherArgs *args = (PhilosopherArgs *)arg;
    int id = args->id;
    int left = id;                     // índice del tenedor izquierdo
    int right = (id + 1) % num_philosophers; // índice del tenedor derecho

    for (int i = 0; i < cycles_per_philosopher; i++) {
        think(id);

        // Solicitar permiso al camarero (semáforo). Solo num_philosophers-1 pueden tomar en conjunto.
        sem_wait(&waiter);

        // Tomar tenedores: primero el de menor índice (para mantener orden y evitar deadlock)
        if (left < right) {
            pthread_mutex_lock(&forks[left]);
            pthread_mutex_lock(&forks[right]);
        } else {
            pthread_mutex_lock(&forks[right]);
            pthread_mutex_lock(&forks[left]);
        }

        // Ahora come
        eat(id, i);

        // Dejar tenedores
        pthread_mutex_unlock(&forks[left]);
        pthread_mutex_unlock(&forks[right]);

        // Liberar espacio en el camarero
        sem_post(&waiter);
    }

    printf("[Filósofo %d] Terminó todos sus ciclos.\n", id);
    return NULL;
}

int main(int argc, char *argv[]) {
    if (argc != 3) {
        fprintf(stderr,
                "Uso: %s <num_philosophers> <num_ciclos_por_filosofo>\n",
                argv[0]);
        exit(EXIT_FAILURE);
    }

    num_philosophers = atoi(argv[1]);
    cycles_per_philosopher = atoi(argv[2]);

    srand(time(NULL));

    forks = malloc(sizeof(pthread_mutex_t) * num_philosophers);
    for (int i = 0; i < num_philosophers; i++) {
        pthread_mutex_init(&forks[i], NULL);
    }

    // Inicializar semáforo camarero a num_philosophers-1
    sem_init(&waiter, 0, num_philosophers - 1);

    pthread_t phils[num_philosophers];
    PhilosopherArgs args[num_philosophers];

    // Crear hilos filósofos
    for (int i = 0; i < num_philosophers; i++) {
        args[i].id = i;
        if (pthread_create(&phils[i], NULL, philosopher, &args[i]) != 0) {
            perror("pthread_create filósofo");
            exit(EXIT_FAILURE);
        }
    }

    // Esperar a todos los filósofos
    for (int i = 0; i < num_philosophers; i++) {
        pthread_join(phils[i], NULL);
    }

    // Destruir mutexes y semáforo
    for (int i = 0; i < num_philosophers; i++) {
        pthread_mutex_destroy(&forks[i]);
    }
    free(forks);
    sem_destroy(&waiter);

    printf("Todos los filósofos han terminado.\n");
    return 0;
}
